### ISSUE TITLE
About porting to webExtension

### ISSUE THREAD

**Issue by StyMaar** [Quicksaver/Tab-Groups#535]
In your farewell post, @Quicksaver said the following :

> Its [Tab Groups] core functionality and basic workflow probably can be made into a WebExtension, but only after an almost complete rewrite of the code (with some major work done on Firefox's side as well!), and still stripped down of at least some of its features. 

Can you tell us more about this ? I fully understand that you don't want or can't port this addon to work with newer Firefox APIs but since many of us are in love with Tab Groups, some may like to give a try on porting this (at least partially) to WebExtension. 

Can you give us a quick overview of what would be impossible in the ported version ? And what would needed on Firefox's side to make it work ?

**Comment by Quicksaver** [comment]
There's already some things about that in #534, probably a good start for anyone who'd like to give it a shot.

**Comment by ghost** [comment]
I'm seriously considering about taking up this project or restarting from scratch and using this as a reference. I only use Firefox for two reasons- Tab-Groups and the tabs-scrolling-than-bunching when there are a lot of 'em. You may not use it outside of coding but it's my bread and butter in how I use Firefox. I have over 60ish tabs and they need to be organized. XD I'm going to start looking into how feasible it is to do this.

**Comment by maverick74** [comment]
Don't know where to post this so i'll post it in both bugs: @Quicksaver Are you aware of:

 [Extend WebExtensions with custom APIs so more legacy add-ons can be ported](https://bugzilla.mozilla.org/show_bug.cgi?id=1215059) ?

**Comment by mikhoul** [comment]
@Quicksaver @maverick74 @maximstewart @StyMaar 

Some Good documentations even a working (but buggy) addon exist for Chrome with Full Documentation:

Original add-on in Alpha stage: https://github.com/arnaud/chrome-tab-sugar

2 Others Fork more advanced and less buggy so they are usable: 

- https://github.com/GodLesZ/chrome-tab-sugar   Updated in 2015
- https://github.com/dan-corneanu/chrome-tab-sugar Updated in 2014

**COMPLETE documentation about the internal of the addon:** https://github.com/arnaud/chrome-tab-sugar/wiki/Interactions

Discussion where the original author of the addon and the "forkers" discussed of the development of the addon: https://github.com/arnaud/chrome-tab-sugar/issues/18


I hope it could help to revive the project !  :smiley_cat: 



Regards :octocat:

**Comment by Chris2011** [comment]
@mikhoul I have to install it on my own, right? The version at the google chrome store is not the current version?

**Comment by mikhoul** [comment]
@Chris2011   Yes, the version in the store is the first Alpha Version which is "buggy" a little bit   :roll_eyes: 

If you want to try those addon in Chrome, you must build a CRX file (addon file) first to be able to install it and use it, but it's easy to do.

**1-** choose which one you wanna try between the 2 forks:

- https://github.com/GodLesZ/chrome-tab-sugar   Updated in 2015
- https://github.com/dan-corneanu/chrome-tab-sugar Updated in 2014

**2-** Go on the page where is the fork and Download the whole project, it is a zip file: http://i.imgur.com/jDGbNc4.png

**3-** Once you have the whole project on your computer just rename the Archive from MyAddon.**zip** to MyAddon.**CRX**  and you have the addon. :smile_cat: 

**4-** To install it in Chrome it's easy too, just follow those instructions: http://chrome-extension-downloader.com/#install

That's it !   :octocat:

**Comment by Chris2011** [comment]
Thx, I will try. A release would be good anyway for all others and maybe a version 2 or so, to bring it into the chrome store.

**Comment by stephen304** [comment]
I hope the chrome versions can have some development activity. I switch between firefox and chrome frequently and I'm constantly disappointed at how all of the chrome tab managers are basically glorified bookmark organizers (mainly how if a tab isn't visible, it means the extension closed it and just saved the url).

**Comment by Nerothos** [comment]
I don't know if I'm late or off topic (if that's the case, please don't mind) but has anyone seen the firefox container? https://testpilot.firefox.com/experiments/containers . Seem like the groups will come back.

**Comment by mechalynx** [comment]
@Nerothos Not sure what you mean. I've already used Containers and I'm not sure its implementation helps at all with restoring functionality to Tab Groups.

**Comment by klint** [comment]
@Nerothos I agree with you that the concepts of Tab Grouping and Containers would fit quite well together at least in my use cases.
I actually use Groups to differentiate various roles I have in life and I bear when navigating in a browser session: work stuff, personal regular stuff, personal dev stuff, personal critical stuff like banking, stuffs related to various associations I'm member or accountant of (for instance). 
Containers will add even better support for that by containing/securing session data but as a side effect,  allow to connect on different accounts to same site in different groups (each corresponding to a role of mine).
The problem I see with containers today are the ease of use: how to display tabs in a given container and hide the others, how to switch the display to another container... Wait, that's exactly what Tab Groups are meant for.
With the same kind of automation: some URLs will be linked to a group/container. Other URLs will be shared accord group/containers (but with different sessions). Some URLs will require their own container and will not share it with any other URL, like banking sites.

@MechaLynx Containers will not help at restoring Tab Groups per se, but they are yet another motivation for Tab Group not to die :)

**Comment by stephen304** [comment]
@klint Perhaps something can be built on top of containers? The only difference between containers and what I need for my use case is having one container "selected", causing all other tabs to be invisible (but still loaded in memory and running). It seems like containers already accomplishes much of the nitty gritty of tab groups by handling container membership and a bonus of isolation.

If it's possible to manipulate containers with an extension, my vote would be to do it that way.

I still have it in muscle memory to press ctrl + shift + e so I'd love to be able to be able to return to that workflow of pressing that shortcut and using arrows and enter to switch groups.

**Comment by Nerothos** [comment]
I was just wondering if container could "replace" TabGroups. I didn't had time to test it so I don't know what features are missing. Sorry if my comment was not useful.

**Comment by stephen304** [comment]
@Nerothos Like in my previous comment, the only thing I think containers is missing is the ability to hide other tabs. I haven't tried containers recently, but last time I checked, it still showed all the tabs in the tab bar. I think it's really close though.

**Comment by nd-** [comment]
@stephen304 I just tried Containers Experiment 2.0.0, and it's possible to show/hide all tabs from a container.
But it's not possible to hide other containers, you have to show/hide them one by one, and the UI isn't very convenient : click "Containers" button in the tool bar, click the right arrow next to the container name, click show (or hide).

**Comment by nd-** [comment]
See https://discourse.mozilla-community.org/c/test-pilot/containers to discuss Containers, especially this post:
https://discourse.mozilla-community.org/t/could-panorama-feature-be-brought-back-to-firefox/14672

**Comment by mirh** [comment]
Which leads to mozilla/testpilot-containers#336

**Comment by nd-** [comment]
> we're definitely watching the 👍 and comments on those issues to help us decide what enhancements to make to Containers in the future.

Time to send your 👍 and spread the word.

**Comment by mirh** [comment]
New bug to follow closely
https://bugzilla.mozilla.org/show_bug.cgi?id=1333837

**Comment by pshem** [comment]
Upvote this: https://github.com/mozilla/testpilot-containers/issues/336
An issue to add tab groups into containers, currently a Stretch goal

**Comment by cpatti97100** [comment]
please if anyone starts a new repo to make tab groups with webextensions, link it here. 
I am thinking about doing just that but hard to know when :( otherwise I would gladly join the work of someone else :)

**Comment by klonamy** [comment]
Any new news about this?

**Comment by mechalynx** [comment]
In the process of figuring out a different issue, I came across these bugs:

 - https://bugzilla.mozilla.org/show_bug.cgi?id=906076
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1333837

There is explicit mention of Tab Groups in the first one. It seems to be a different extension, but the second bug seems explicitly involved with creating tab groups under WebExtensions.

Related comments:

 - https://bugzilla.mozilla.org/show_bug.cgi?id=906076#c234
   > The irony of this is that, as I see it, the main benefit of this change is that it will make it easier to use large numbers of tabs.  But it looks like the fix is not going to be complete until after XUL extensions are dropped, so the main extensions that are also needed for using large numbers of tabs (such as the very useful Tab Groups Helper written by the main person working on this bug) will no longer be available, and it will still not be possible to comfortably use large numbers of tabs in Firefox.  Is that right or am I missing something?
 - https://bugzilla.mozilla.org/show_bug.cgi?id=906076#c251
   > Wasn't the tab groups addon mentioned before in regards of prematurely inserting lazy browsers?
Was it determined if it's an issue on FF's side or the addon's (like with BackTrack Tab History)?

I haven't read the full details on each of these - apologies if it's a false positive.

**Comment by mikhoul** [comment]
Here.s a WebExt that can be used as replacement of TabsGroups, maybe I will fork it to Firefox and improve it at a later time but it is usable.  :octocat: 

https://chrome.google.com/webstore/detail/divider-tabs/mpbblhoepkmpgalgppdoahgeoknmjgpc/reviews

**Comment by nd-** [comment]
Make it work on firefox and you're a hero !
At last, a new hope... :)

**Comment by Deledrius** [comment]
Not bad.  It doesn't look like it quite supports the same workflow, and leaves things cluttered still, but the alternative is far worse.  

Also, it would be ideal for any replacement to be able to import existing groups.

**Comment by nd-** [comment]
@mikhoul Too bad, it doesn't work when divider tabs are pinned.

**Comment by grahamperrin** [comment]
Divider Tabs 2.0 (2014-10-07): 

- does not work with Firefox 54.0 (64-bit) on FreeBSD, divisions are empty
- is not reliable with Chromium 58.0.3029.110 (64-bit) on FreeBSD.

I can't imagine it as a successor to Tab Groups. Sorry.

**Comment by BenjamenMeyer** [comment]
I have yet to find a Chrome tab manager that works anything like FF's Pandora/TabGroups functionality - even https://github.com/Quicksaver/Tab-Groups/issues/535#issuecomment-312361904 doesn't really seem to do so. This extension really is the prime differentiator for Firefox. Sad to FF going this way, and enables really good tab management and workflows:
- minimal number of tabs in a window
- management "windows"/"tab" where you can manage all the tabs and even go to a specific tab in a specific tab group.
- restores when session crashed

**Comment by kah0922** [comment]
[An API for showing/hiding tabs has been approved by Mozilla.](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515)

**Comment by nd-** [comment]
![starwars_epi4_01-4f17cf46f1dd](https://user-images.githubusercontent.com/5727669/28845837-97ad05ce-7709-11e7-9fa8-6300c5aab288.jpg)

**Comment by mikhoul** [comment]
Here's an Alternative actively developed: https://github.com/denschub/firefox-tabgroups

Regards :octocat:

**Comment by dgutov** [comment]
The alternative is yet to have a WebExtensions port either, though.

**Comment by tetris4** [comment]
Just adding another recent relevant bug report:
https://bugzilla.mozilla.org/show_bug.cgi?id=1384515

**Comment by ghost** [comment]
As an update here, I've started learning how plug-ins are created. This will be slow but I am determined in getting something similar up and running soon. Time is a bit of an issue though. But I have not forgotten about this project.

**Comment by e-motiv** [comment]
I found a Firefox native workaround for Tab Groups and Tab Groups Menu functionality. I'm using bookmarks, bookmark toolbar and -(sub)folders for it.
It's not perfect, but it can also serve as a basic model to make it perfect again. You see, I believe that by using these native techniques (or simulation) one can make a super simple and lightweight webextension with it!

More detailed and illustrated info on it in this article:
http://attic.e-motiv.net/fyi/tab-groups-workaround

I hope this helps...

**Comment by Deledrius** [comment]
Using bookmarks loses the tab's history and potential contents. 🤕  It may be sufficient for some workflows, but not mine.

**Comment by e-motiv** [comment]
Yeah. Though I don't mind and I didn't think of that, I guess that can be a
game changer.

FMI: What exactly did you mean with "potential contents"? Did you mean the
content that could otherwise be reached via history or did you mean an
extra unrelated disadvantage?

Mvg,
[image: Ruby Contact Card]
<http://www.e-motiv.net/Ruben%20Goethals%20-%20e-motiv.vcf> Ruben
<Ruben@e-motiv.net>    @   e-motiv <http://www.e-motiv.net>
[image: Ruby] <http://www.e-motiv.net>
+32  (0) 484  /  23 78 78

**Comment by klint** [comment]
I have found using the Containers experiment together with these 2 add-ons quite effective for me, replacing Tab Groups (basically, I have matched groups with containers).
https://addons.mozilla.org/fr/firefox/addon/sea-containers: sidebar showing all containers with their tabs
https://addons.mozilla.org/fr/firefox/addon/switch-container: allows to switch a tab from one container to the other

**Comment by irvinm** [comment]
> I have found using the Containers experiment together with these 2 add-ons quite effective for me, replacing Tab Groups (basically, I have matched groups with containers).
https://addons.mozilla.org/fr/firefox/addon/sea-containers: sidebar showing all containers with their tabs
https://addons.mozilla.org/fr/firefox/addon/switch-container: allows to switch a tab from one container to the other

This seems like a pretty good entry level solution.  One thing that is really missing is the ability to treat the tabs in a tree but see that has already been entered as a feature enhancement.  (Both TreeStyleTab and "Tree Tabs" are shaping up to be good vertical tab solutions with webextensions.)  If we can get vertical tree behavior along with hiding\unhiding tabs ... we might be on to something!

**Comment by maverick74** [comment]
There is also this, but it's only for FF57+:

[Taborama](https://addons.mozilla.org/en-US/firefox/addon/taborama/)

**Comment by ahjolinna** [comment]
to follow @maverick74 post, taborama was renamed to conex and moved to https://github.com/kesselborn/conex ... see https://github.com/kesselborn/conex/issues/46 for reasons
&
the official addon page  : https://addons.mozilla.org/en-us/firefox/addon/conex/

**Comment by StyMaar** [comment]
> Furthermore, my taborama actually does not have a panorama'ish view (which was where it was derived from), so the name does not make sense at all

Then it's not an alternative to tab-group, is it ?

**Comment by ahjolinna** [comment]
@StyMaar : well, there wont be a direct replacement until this happens: https://bugzilla.mozilla.org/show_bug.cgi?id=1384515 ...until then this more or less the best what we can get as an replacement

same goes with this project: https://github.com/denschub/firefox-tabgroups/issues/60#issuecomment-321878075

**Comment by dgutov** [comment]
Meanwhile, when I enabled legacy extensions in Firefox 57 and tried it, Tab-Groups was one of the few legacy ones that still appeared to function correctly. Had to downgrade to 56 because of the other broken ones 😞

**Comment by mikhoul** [comment]
>  Had to downgrade to 56 because of the other broken 

@dgutov  Which one were broken from 56 to 57 ? 

I have 56 right now and I was fearing that many of my addon will broke if I upgrade to 57 even with legacy addon enabled ans you seem to confirm my fear. 😞

**Comment by klint** [comment]
Downgrading is now not supported anymore, officially at least. Was it seamless for you? ☺️

**Comment by mechalynx** [comment]
I downgraded and it was pretty smooth, all I had to fix was my menu items. Supported or not, here's links to getting an installed for a previous version, for those interested:

Firefox Release & Beta: https://download-installer.cdn.mozilla.net/pub/firefox/releases/
Firefox Developer Edition: https://download-installer.cdn.mozilla.net/pub/devedition/releases/
Firefox Nightly: https://download-installer.cdn.mozilla.net/pub/firefox/nightly/

Of course, back up your profile before downgrading.

**Comment by dgutov** [comment]
> Which one were broken from 56 to 57 ?

Pretty much all the rest, IIRC. I'm pining for Undo Close Tab Replacement, Old Reader Notifier, Tab Wheel Scroll and KeeFox (this one has a WebExtension beta, fortunately).

**Comment by Spitfire1900** [comment]
Just using ESR for as long as possible over here.

**Comment by grahamperrin** [comment]
> … not an alternative to tab-group, is it ?

It's good for some use cases. See _Finding tabs across multiple groups (Tab Groups) and multiple windows_ under https://discourse.mozilla.org/t/-/17087/60?u=grahamperrin

**Comment by klint** [comment]
The new Firefox Container add-on now offers a way to hide&show a Container, ie all tabs belonging to a certain container. Starting with FF57.
It still lacks some nice UI and buttons (this could be provided by another add-on?) but I think we might have something quite interesting there.

**Comment by grahamperrin** [comment]
Whilst there's much to love about Firefox containers – without and with extensions – there's an emphasis on containment **without movement**. 

The seamless quality of movement that is associated with _groups_ (e.g. moving from one group to another) is not possible with _containers_.

**Comment by Deledrius** [comment]
I find it very disappointing that Mozilla has been so actively hostile to enabling a workflow that was a part of the core browser for five years.

For me, anything that doesn't migrate my existing tab groups and allow me to continue using them mostly as-is is a non-starter, and I'll be stuck on ESR until the API catches up.

**Comment by kesselborn** [comment]
@StyMaar is the panoramic view one of the major feature that you use of tabgroups?
Basically, I write conex to have a decent tabgroups replacement and implement the freatures I use most first (ego-driven-development). Technically there is no reason for the panoramic view is not there -- feel free to open an issue at [conex](http://github.com/kesselborn/conex).

@Quicksaver thanks again for tabgroups: it's what kept me at firefox all the time and it's the reason why I try to create something that tries to come close to tabgroups.

**Comment by ghost** [comment]
As an update, while I was thinking of ways the lack of proper Tab API in WebExtensions, one option that came to mind is hiding the original tabs bar and create a new one based on the data there in. They could be made into buttons or something similar that has options for hiding/closing, etc. It isn't performant as it might take a lot of memory to duplicate icon info and generate alternative "buttons" but could work for the time being until or if the latter option is made available. This is all based on being able to hide and create bars though and I'm not certain yet if such an option is present. I know it used to be a thing since I had a plug-in that could hide all bars but I don't know much about the new extension api.

**Comment by kesselborn** [comment]
@maximstewart that is currently not possible

- hiding the tab bar is not possible via an extension (only via `userchrome.css`) -- I think it'll be part of the [show / hide tabs api](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515) ... but if that is implemented, your described work around is hopefully not necessary anymore
- creating an addon-toolbar [is not possible at the moment either](https://bugzilla.mozilla.org/show_bug.cgi?id=1215064)

the [show / hide tabs api](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515) is making progress, however: currently most likely hiding tabs will mean that they will be discarded as well, which is not feasible for a tab group replacement :( ... discussions went a little bit hot on that bug ;) 

The first implementation of that api was different and worked quite well with conex: https://www.youtube.com/watch?v=wTwmIFSnLWY ... let's see how the api will develop

**Comment by Deledrius** [comment]
It's been very frustrating watching the (lack of) progress on [that bugzilla item](https://bugzilla.mozilla.org/show_bug.cgi?id=1384515), seeing it occasionally take one step forward, and then two diagonally back every few weeks as the final deadline creeps ever closer.

Now, with some of the latest comments, it seems more clear that all of this delaying isn't _solely_ due to incompetence or lack of interest, but possibly a personal vendetta against the entire idea of Panorama and allowing users to have a Tab Groups extension by someone claiming executive power over the API.  Very disappointing.

**Comment by grahamperrin** [comment]
In conversation below <https://mozilla.logbot.info/firefox/20171026#c13749315>: 

>> … a straight up rewrite/replacement in the works …

>> …  from a Mozilla developer (or multiple ones) but I do not think it is an official Mozilla project

> [if/when details become public I'll be glad to know about it. Thanks](https://mozilla.logbot.info/firefox/20171026#c13749596)

**Comment by Deledrius** [comment]
"how often, if ever, do you move a page/tab from one group to another?"
All the time.  It's quite a common occurrence that a tab is opened in my current group, but because that's the wrong context for the new tab, I move it (or several, thanks to [Multiple Tab Handler](https://addons.mozilla.org/en-us/firefox/addon/multiple-tab-handler/)) to the correct group.

I use tab groups as a compartmentalization of tasks (different open freelance jobs I'm working on), and to group similar things (multiple audio tabs, such as YouTube).  I expect to be able to switch between them on demand, I don't expect or desire them to be sandboxed from each other (I may have github opened and logged in a tab in each group, for instance), and I expect them to keep running when I send them to a group which isn't currently the active one (as I want to listen to a YouTube video but not want it cluttering my current task's tab space).

This seems obvious and trivial, but the questions I see from Mozilla in Bugzilla makes me wonder if anyone there even used a feature that was core in their browser for several years (of course, given how many times they've tried to kill it, I think I know the answer to that).  Even the people who are trying to support the features needed by tab groups (and I deeply appreciate their efforts) don't seem to grasp *why* or *how* it's used.

I'm stuck on ESR, because I have a large number of extensions which are being deprecated with no equivalent replacements.  I'm a developer myself, but I can't rewrite them all, and with the attitude I see coming from the browser's developers when being asked for simple API enhancements to make things possible, it doesn't make me eager to spend my time on dead-ends either.  If you aren't a top-10 extension, no one cares.  The new extension API model means you have to request permission from the developers before you can implement something, which is a huge shift that kills the long tail ecosystem Firefox has enjoyed since the very beginning.

Firefox used to pride itself on being customizable to the needs of the user.  Ever since around Firefox 4, there has been an increasingly hostile attitude against anyone using it outside of the way Mozilla wants to dictate it be used, and I'm not even sure they realize it.  They seem baffled and unresponsive to the outcry every time they take away another degree of flexibility.

**Comment by grahamperrin** [comment]
Briefly: 

> … wonder if anyone there even used a feature that was core in their browser …

– _certainly_, Tab Groups was used by the former Firefox Engineering Manager (desktop). Beyond that, I can _reasonably assume_ that the technology was/is **highly valued by numerous people at (and around) Mozilla** both (a) whilst it was integral to the browser; and (b) after it became an extension. 

> … stuck on ESR, … 

I became happy after making the simple decision to continue with 56.x until maybe August 2018. The lapse in support will not bother me; I'm accustomed to mixing multiple bleeding edge environments in which **Tab Groups 2.1.4 is solid as a rock** – https://twitter.com/grahamperrin/status/925522450410037249 – with more than a thousand tabs and eighty-something groups. 

Less selfishly: I expect to throw myself wholeheartedly into thorough testing of whatever extension (or combination of extensions) will succeed 2.1.4 for users of a future release of Firefox. 

----

Whilst the current situation is not ideal, we can not change history. For now: please be kind to developers. I stumbled across this topic whilst browsing some of the [community-oriented badges](https://discourse.mozilla.org/badges#ember1181) at Mozilla Discourse: 

- [Preventing burnout](https://discourse.mozilla.org/t/-/2944?u=grahamperrin) …

**Comment by Deledrius** [comment]
Sorry, I didn't mean for that response to turn into a bit of a rant, but this whole situation (both for Tab Groups and WebExtensions as a whole) has been incredibly frustrating, more for how it's been handled than the justifiable reasons for the change.

> Tab Groups was used by the former Firefox Engineering Manager (desktop). Beyond that, I can reasonably assume that the technology was/is highly valued by numerous people at (and around) Mozilla both (a) whilst it was integral to the browser; and (b) after it became an extension.

I'm glad to hear it, but it also surprises me because I see so little support for the notion when browsing the last year of public issues and requests on the topic.  I don't know enough of the internals to do more than make observations on how it looks *from the outside*.  Are any of those supporters still at Mozilla?

> I became happy after making the simple decision to continue with 56.x until maybe August 2018.

I've done that on a non-primary installation, and you're correct that it currently works great.  But I made the decision to stick with ESR when 53 was released because at the time I wasn't sure how soon they'd break things prematurely (and there were already changes in Nightly that were introducing blockages to rolling back profiles).  I happily used the Developer Edition and Beta Channel as my primary installations until this decision.

**Comment by BenjamenMeyer** [comment]
FWIW - I have lots of tabs and tab groups, and like the hidden+active state, and really part of what mitigated that was that Firefox didn't load a tab until you clicked on (per configuration) so not all tabs were always loaded in all groups - just those I activated. Never had a performance issue due to that, and run-away tabs were brought to attention and could be stopped.

**Comment by Deledrius** [comment]
Same here.  Once Firefox got the lazy-loading of tabs on startup this was never a problem, and if needed, a restart of the browser (instigated by the user for this express purpose) would put everything back in a dormant state.  It seems obvious to me just through basic usage alone that whether a tab is loaded has nothing to do with whether it's in an active group, and that it's a detail best left as an option to the extension.  

Discarding tabs as an inherent part of hiding them is a premature optimization that breaks existing workflows.

**Comment by grahamperrin** [comment]
From what I recall of the difference between 54.x and 55: 

- the time 'saved' when starting Firefox is easily lost, if you count the sum of the waiting periods for the group management view for each window

– I currently have 1,296 tabs across fourteen windows, with one of those windows given to 419 tabs, so YMMV. 

Side note: 

- realistically, fourteen = thirteen because I leave the first window open but avoid using it
- see [post 34](https://bugzilla.mozilla.org/show_bug.cgi?id=1378459#c34) under Mozilla bug _1378459 - Allow some addon functionality to load prior to any content loading_.

Here's a screenshot, taken yesterday, of a window that now has 300 tabs: 

![2017-11-03 07 58 48](https://user-images.githubusercontent.com/192271/32403254-2b1ba4ba-c12d-11e7-85a2-14f0c2d4ce6f.png)

– with two notable exceptions ('mash' and questionable'), each group focuses on a single extension.

**Comment by ghost** [comment]
Just an update... Here is something minor and kind of gets at the idea. @kesselborn  Has mentioned that my former idea about hiding the tab bar wont work as there is no tab bar hide functionality. He did mention the show hide functionality api being worked on which I am aware of but wasn't waiting on to try 'n get something cobbled together. Jesus Firefox, ya'll done goofed badly. Still have hope though that what I currently have posted might pan out....

Here is my Twitter post with an image of what is made.
https://twitter.com/ITDominator/status/929598222275874816

I'll be looking at @kesselborn 's link of : https://www.youtube.com/watch?v=wTwmIFSnLWY

Please note that my attempts are at a very early stage and time for me is scarce to work on this. But I can not move to the new FF until this is resolved.

**Comment by Keith94** [comment]
New "Simple Tab Groups" WebExtension by @drive4ik just released here: https://addons.mozilla.org/addon/simple-tab-groups/ :)

**Comment by JPFrancoia** [comment]
@Keith94 yeah, "Simple tab groups" looks better than https://github.com/Morikko/sync-tab-groups, the replacement I'm using right now. Of course it's not quite the same thing as our godd ol' tab groups: history is not saved, tabs are unloaded when changing groups, and there is no drag n' drop interface.

**Comment by Keith94** [comment]
@JPFrancoia I made an issue for [keeping tabs loaded](https://github.com/Drive4ik/simple-tab-groups/issues/2), so keep your fingers crossed for that. ;)

**Comment by OndrejNecas** [comment]
@Keith94 Is it possible to import tab group from Tab Groups to Simple Tab Groups? (via JSON file or something)

**Comment by mirh** [comment]
@OndrejNecas You can do that in [conex](https://github.com/kesselborn/conex).

**Comment by Drive4ik** [comment]
> Is it possible to import tab group from Tab Groups to Simple Tab Groups? (via JSON file or something)

@OndrejNecas
Yes. I am working on this functional

**Comment by kesselborn** [comment]
@Drive4ik although that is part of the uglier code in conex, here is how I read the backup file:

https://github.com/kesselborn/conex/blob/f178ff34d8c3488b653fa59ec36f276afee69b4a/conex-options-ui.js#L4-L49

It dumps the backup file to groups with their respective tabs ... perhaps it helps

**Comment by PX03AFK** [comment]
I have just been updated to FF 57.  I have been using tab groups for a long time and my immediate reaction to containers is that the idea sounds great but without tab group type functionality it is not at all user friendly.  I don't want visibility of all my tabs, why else would I want to group them.  So, I don't know if I'm missing something with the containers but it appears I can group things but have to have multiple clicks to hide a container and when I do click on hide, nothing happens.

**Comment by BenjamenMeyer** [comment]
@Drive4ik looking nice; please keep us up-to-date with your progress.

I've locked my FF install to FFv56 and am not going to upgrade to FFv57 until I have a migration path. Yours looks promising. Do you have a screen shot of a tab manager to manage the groups?

**Comment by Drive4ik** [comment]
@BenjamenMeyer you can "install" [Mozilla Firefox 57, Portable Edition](https://portableapps.com/apps/internet/firefox_portable) and try install https://addons.mozilla.org/ru/firefox/addon/simple-tab-groups/ in portable version, after that you can import backup of old "Tab Groups" addon into Simple Tab Groups (STG) and check for operability.

But STG has a issue - reload tabs after change group. I'm looking for a solution to this problem.

Addon page https://github.com/drive4ik/simple-tab-groups

Now I'm working on "Move tab to group" functionality.
Please, wait few days for the new version.

**Comment by afranke** [comment]
> after that you can import backup of old "Tab Groups" addon into Simple Tab Groups

How exactly do you do that backup?

**Comment by Drive4ik** [comment]
@afranke 
Go to about:tabgroups#session and click "Create Backup File"

**Comment by PX03AFK** [comment]
It appears that Simple Tab Groups is incompatible with containers.  I just enabled it and it moved all my tabs into the default.  Given that containers is what seems to be the way FF has gone, it would be useful if Simple Tab Groups at least left the containers alone or maybe even used them.

**Comment by kesselborn** [comment]
@PX03AFK shameless self plug: [conex](https://github.com/kesselborn/conex) uses containers for grouping which some people hate and some people like ... perhaps it fits your needs

**Comment by PX03AFK** [comment]
The grouping with containers isn't working in the same way as Tab Groups, at least not at present.  With the old tab groups the only tabs visible were tabs in the group whereas with containers there doesn't, as yet, seem to be a way of showing just tabs in a container, all the tabs are visible.
Shame really, as I have liked FF for years but I'm now trying Waterfox as it still provides the tab group functionality that I wish to use.
Sorry to say this but this change to FF is somewhat like Microsoft when they ignore what people like and simply change it because they think it's better.  I realise that supporting older capabilities might be an overhead but surely it would be worth keeping users happy rather than sending them off to find alternatives.

**Comment by mirh** [comment]
> whereas with containers there doesn't, as yet, seem to be a way of showing just tabs in a container, all the tabs are visible.

https://github.com/kesselborn/conex#getting-a-version-that-can-really-hide-tabs

**Comment by PX03AFK** [comment]
I know I didn't try for very long with the new FF but I was frustrated with the complete change of functionality and so I've switched to Waterfox as it still uses the same tab groups capabilities I like.  If containers ever get to be similar I will try but until then I will stop using FF.

**Comment by cpatti97100** [comment]
I share your concern on the topic since I've been a heavy user of this plugin. Software goes on anyway soon or later and I moved to 57 with great benefits.
Now I would like to help to develop a replacement but I understand there's a fundamental API missing, the one for showing\hiding tabs, or is it doable anyway? @Drive4ik do you need that to move your plugin forward? It seems like it is the best attempt at a replacement so far

**Comment by Drive4ik** [comment]
@clivend 
I read all api and did not find a way to create a tab without loading it 😕 
This is bad news, but maybe mozilla will add it in the future.
We will wait and be given

P.S.
STG updated to v1.7
Added functional to move tab to group

**Comment by cpatti97100** [comment]
just have to keep an eye on this
[https://wiki.mozilla.org/WebExtensions/TabHiding](https://wiki.mozilla.org/WebExtensions/TabHiding)
[https://www.dropbox.com/s/b58qn5vvuqzokf0/tab-hiding-movie.mov?dl=0](https://www.dropbox.com/s/b58qn5vvuqzokf0/tab-hiding-movie.mov?dl=0)

as soon as it lands, we'll have our beloved plugin back :)

EDIT fixed links sorry

**Comment by JPFrancoia** [comment]
These two links are dead. Can we have them back please ?

**Comment by mirh** [comment]
Actually working
https://wiki.mozilla.org/WebExtensions/TabHiding
https://www.dropbox.com/s/b58qn5vvuqzokf0/tab-hiding-movie.mov?dl=0

**Comment by BenjamenMeyer** [comment]
@mirh now if only there was some kind of schedule...waiting to update FF until that kind of functionality works and extensions are ported. In the mean time, I'm primarily using Chrome, but I am *really* missing TabGroups.

**Comment by Nerothos** [comment]
If you really want to use "firefox" then in the meantime you can use Waterfox for exemple. It's still support this addons.

**Comment by ghost** [comment]
OK, as updates go, I have nothing too new. The big thing is I am getting very comfortable with creating plug-ins.I've created 3 or so minor things but each affects tabs and the way I use them. Anywho, one interesting thought came to mind. That is the idea that one need not hide tabs to get a working groups system. While having the bar filled with tabs is unsightly, there can be a system that is separate from that by way of the browser_action window.  One would just need to get used to using that and ignore for the time being the tabs bar.

At this point, I have that minor example in the previous post. With the knowledge I have gained through creating the other plug-ins, I think I am ready to start tackling the task of generating something actually useful and working. I must once more re-iterate that I do not guarantee progress but I have been lucky so far in the amount of time I have had to consider this. The next post will either be one of progress or admittance of defeat. Wish me luck. =]

**Comment by Drive4ik** [comment]
Hi @maximstewart 

You mean here is behavior?
![stg_windows_behavior](https://user-images.githubusercontent.com/7843031/33828871-806ccf6e-de76-11e7-9a0e-1ddfb97e878d.gif)

**Comment by ghost** [comment]
@Drive4ik  Not really. I'm going for the approximate grid style as the original tab groups. A list doesn't give me the info I need since I'm more top down and visual grid oriented kinda person. The option you showed is great but not what I want and need. <shrugs> I'm so glad I know how to code because I'da really lost my Sh** if I couldn't. This situation sucks but I can fix it to meet my needs given time.

**Comment by BenjamenMeyer** [comment]
@maximstewart personally, I like the list for quickly switching, but I also like the expanded UI for managing the groups.

**Comment by JPFrancoia** [comment]
TbaHide recently landed in nightly: https://bugzilla.mozilla.org/show_bug.cgi?id=1423725

**Comment by stephen304** [comment]
TabHide has landed in 59 beta: https://blog.mozilla.org/addons/2018/01/26/extensions-firefox-59/

Hidden behind `extensions.webextensions.tabhide.enabled`

**Comment by Drive4ik** [comment]
Hi all!
I make [EXPERIMENTAL] Simple Tab Groups with support Hide Tabs API:
https://addons.mozilla.org/firefox/addon/experimental-simple-tab-groups/
It work on latest FF Nightly and Beta.
To make it works, visit `about:config` and set `extensions.webextensions.tabhide.enabled` to `true`.

Please, read
https://github.com/Drive4ik/simple-tab-groups/issues/2
Please tell me - what difficulties / bugs do you experience with this addon.

Thanks.

**Comment by klint** [comment]
Thank you :)
I will test and let you know
Cheers

2018-03-06 17:21 GMT+01:00 Drive4ik <notifications@github.com>:

**Comment by stephen304** [comment]
@Drive4ik Great work! I've installed it as I'm anxious to get away from tree style tabs (doesn't work well for my workflow). I hope you don't mind all the issues I'm opening - I have a lot of feedback!

**Comment by Drive4ik** [comment]
@stephen304 Hi! Thanks for feedback. I will try to make it even better)

**Comment by ghost** [comment]
OK, it's update time. I've not got grouping yet but I do have a solid base to work from. Presently, the big thing is I can search through my tabs and since I'm only ever dealing with a small subset it isn't near as hard to recall what I'm searching for. While the plugin's main functionality is searching, it's made better by way of the grid representation and the icon features such as closing, creating new tabs, and re-ordering through drag-n-drop actions. These features will be used when I implement grouping which is why I did all this before the grouping.

Now, it's worth mentioning that I had grouping working but it was too complex and so I scratched it to start from a solid base and branch out. This is why I have the current plugin the way it is. I am using dragula.min.js to do the heavy lifting as is and I intend to expand upon it  for the grouping now that many of the features that I need are present.

So here is the thing, the browser_action window is pretty tiny and it makes it harder to create groups and manage/move them around. In addition, since it's through that, saving the "session" has to happen every time the window closes and needs to load it back in each time the window opens. This is pretty inefficient or poses issues at the least when there is a lot of data to load/unload. I'm thinking about how to handle this but there are no clear answers that make it lightweight and resource efferent.

[  My current idea for the data is as follows  ]
WindowID:
    GroupID:
        TabFaveIco | TabsURL | TabID

One selects the window or loads a default one and it loads the group ids into their respective HTML containers. The containers then populate with their URLs, ICOs, etc. Here is where I'm thinking I can make it more efficient. Rather than load all containers, again, a default container loads. When you want to load another container or create one, then the container loads and the other unloads unless one is moving stuff around. This means that not all the data has to load/unload at once. Anywho, that's where I'm at at this point. Defeat has yet to claim me but it is slow going.


@Drive4ik By the way, I'm sorry if my reply was terse as I didn't mean for it to seem like that. I think your plugin is very cool but it just isn't what I'm looking for. I didn't intend for it to sound like I was poopooing on you efforts. =P I'm sorry if it came across that way as when I looked back it seemed clipped.

**Comment by kesselborn** [comment]
@maximstewart if the popup space is a constraint, you can just open a new tab and programmatically create the html code for it (just as you do with the browser-action-popup) ... `browser.runtime.onConnect` is your friend here -- if you need a pointer to working code, feel free to ping me -- it's a little bit tricky.

Regarding data size: if you save it as json, the size should be no problem. You probably want to save the data for every tab-create and tab-close event instead of only window-close ... just in case the browser crashes.

If you want to avoid the hassle of implementing the state of grouping, you can rely on conextualIdentities aka containers so Firefox handles all the state for you -- however: people will criticize you a lot for it as some have the opinion, that containers should not be used for grouping but for isolation only (which is an opinion you might disagree with as well).

Happy hacking

**Comment by grahamperrin** [comment]
Recently I began using [Tip Tab](https://addons.mozilla.org/addon/tip-tab/) with [PopUp](https://addons.mozilla.org/addon/popup/) (and Conex, and legacy Tab Groups): 

![2018-05-06 11 36 19 conex tip tab and popup with waterfox](https://user-images.githubusercontent.com/192271/39715483-56a7ee3e-5225-11e8-8a4f-00405504c515.png)

– the Tip Tab tab is popped up to a window, which I maximise on the display to the right of my notebook. 

@kesselborn this is comparable to the use case that I envisaged in [a bundled web page for Conex · Issue #52 · kesselborn/conex](https://github.com/kesselborn/conex/issues/52).

**Comment by ghost** [comment]
@kesselborn 

"...if the popup space is a constraint, you can just open a new tab and programmatically create the html code for it (just as you do with the browser-action-popup) ...browser.runtime.onConnect is your friend here."

Hey, that would work! The only thing that makes me hesitant is that this makes me think it affects new tabs. Is this the case? If not, then would you send me some pointers about how you'd set it up

"Regarding data size: if you save it as json, the size should be no problem."

I was thinking much the same and it might be easier to insert/remove tab instances.

"You probably want to save the data for every tab-create and tab-close event instead of only window-close ... just in case the browser crashes."

When I said window I was more referencing the pop-up window for browser_action. XD So, if it closes which is after focus is lost then it writes the new data.

"If you want to avoid the hassle of implementing the state of grouping, you can rely on conextualIdentities aka containers so Firefox handles all the state for you -- however: people will criticize you a lot for it as some have the opinion, that containers should not be used for grouping but for isolation only (which is an opinion you might disagree with as well)."

I'm in your camp of thinking and my early thoughts were around this method. But, it then does undermine the idea of using identities as a means of one being work, home, other such that cookies and session data follows the instance. XD In addition, working with data like this is a good exercise for a noob like me so I'm forging this option. =]

**Comment by grahamperrin** [comment]
> … [a fundamental API](https://github.com/Quicksaver/Tab-Groups/issues/535#issuecomment-346660381) …

Catching up here … there seems to be no **groups** API. Am I missing something?

From [WebExtensions API for groups (for tab groups) - Add-ons - Mozilla Discourse](https://discourse.mozilla.org/t/-/28533?u=grahamperrin)

> … pockets of brilliance, but the **overview seems to lack cohesion**, …

Thoughts? 

TIA

**Comment by maximstewart** [comment]
OK...I have a grouping plugin working but currently there is a lot of jenk. Recall that this is based off of my other plugin : https://addons.mozilla.org/en-US/firefox/addon/tab-search-smarter/
I have split from the other plugin to have a code base that is solely for the groupimg stuff while the other remains strictly an nicer search plugin.

https://github.com/maximstewart/Tab-Groups/

NOTE:  Get the current test build through the testBuilds folder.

Current jenk issues:
1. This can only work in a single window instance else new window will overwright the former session OR not see its own session. . (This isn't jenk per say since I just haven't implemented a window id check and pull method.)
2. Saving occurs on an onmoveout event and it fires or can fire a lot. This is bad b/c if the pop-up window closes it can screw up the save if it's saving when the window closes.Thus, the loading of tabs is messed up. Pop-up windows seem to get killed brutally without waiting for tasks to complete.... T.T
4. Containers can drag sometimes which then causes nested HTML containers. (Might have fixed this but it's not in the test build.)
5. The pop-up window is small. It can be difficult to move tabs between containers. (This is fixable by doing @kesselborn 's suggestion of a "new tab and programmatically create the html code". )

**Comment by grahamperrin** [comment]
## Multiple windows

Re: https://discourse.mozilla.org/t/-/28533/6?u=grahamperrin I'll want some groups to **span multiple windows**. (To a degree this probably comes naturally to Conex, which is a great extension, but groups should complement containers). 

Possible?

## Interoperability

I'll need any group that is created by (say) Panorama View to be **also** manageable by (say) Tiled Tab Groups; and vice versa. Extensions should work with each other, without requiring a per-extension set of groups. 

Possible?

TIA

**Comment by grahamperrin** [comment]
From https://github.com/denschub/firefox-tabgroups/issues/60

> … Conex … additional functionality of integrating with … Containers 

@ScottRFrost it's an **excellent** extension, however I'm not sure about the word _integrating_ with containers. 

There's unavoidable _contention_ between some of the goals of the technologies.

**Comment by BenjamenMeyer** [comment]
### So has is there a general consensus on a good replacement yet?

I don't have a system I'm willing to try out upgrading Firefox and playing with extensions on - don't want to lose my existing tab groups. But I haven't seen anything striking yet in the comments, and Chrome is starting to drive me nuts with its resource consumption for the number of tabs I end-up keeping open.

**Comment by Pab0** [comment]
@BenjamenMeyer I've been using [Simple Tab Groups](https://addons.mozilla.org/en-US/firefox/addon/simple-tab-groups/) for a couple months now, and while not on the same level as the original, it does its job. It has search, pinned tabs stay always visible, and the UI is quite similar. I can't comment on the container or multiple window handling, since I'm using neither.

The drawbacks so far are:
- When switching to another group, all tabs in the new group are loaded (not on browser startup though).
- When switching to another group, all tabs in the old group are unloaded.
- In two or three occasions, I mistyped a key combination which removed all unloaded groups. The first time I lost everything, but the other two times I just closed and reopened the browser and everything was there. The group losing issue might have been resolved by now; the first incident happened months ago.

Not to discredit any other alternatives - this just happens to be the only extension I've tried so far. I'd be interested in hearing how the rest of the alternatives are progressing.

**Comment by klint** [comment]
The Experimental version of Simple Tab Groups now hide tabs instead of unloading them.
Still a few bugs, but gets better and better :)

**Comment by mikhoul** [comment]
@BenjamenMeyer   You can easily migrate/import your existing groups without losing anything in [Sync-Tab-Group](https://morikko.github.io/synctabgroups/)  in Chromium you can also [The Great Discarder](https://chrome.google.com/webstore/detail/the-great-discarder/jlipbpadkjcklpeiajndiijbeieicbdh) to lower your memory usage.

Right now in my active group I have more than 150 tabs with no problem at all. :smile: 

Regards :octocat:

**Comment by BenjamenMeyer** [comment]
@Pab0 thanks for the info; seems like if they solve the unloading issue it may be doable so may be next release (re @klint 's statement about the experimental, assuming everything goes to release)?

@mikhoul Chromium isn't quite an option unfortunately so it's Chrome and Firefox and Firefox has long been my primary choice in many respects because of the Pandora/TabGroups functionality. Chrome/Chromium doesn't have anything that competes - I've tried to find something, it's just not there. So really looking forward to having a new Firefox tool to do the same thing as Pandora/TabGroups.

**Comment by ghost** [comment]
As an update, I've decided to just stick with what I've created and offload my groups into other things like bookmarks and/or notes/databases. IE, i've gone through my stuff and determined if its worth keeping and if so moved it to bookmarks or one of my databases for future reference. The plugin I created is 

Tab Search & Manage:  https://addons.mozilla.org/en-US/firefox/addon/tab-search-and-manage/

I can search through tab titles, hide them, create new ones, and other stuff. It meets my current needs but has no grouping.

**Comment by CubeTheThird** [comment]
It seems these days there are several tab grouping addons popping up. At a glance, the one that's peaking my attention is [Panorama Tab Groups](https://addons.mozilla.org/en-US/firefox/addon/panorama-tab-groups/). The interface seems to be very similar to what this has previously offered, not to mention the development is active (many addons haven't been updated in many months). I haven't spent much time looking at it, but it's one I'm excited to test out.

**Comment by Deledrius** [comment]
@CubeTheThird: That's the one I've been using, and so far it seems the best replacement.

**Comment by klint** [comment]
I've been using [Simple Tab Groups](https://addons.mozilla.org/en-US/firefox/addon/simple-tab-groups)  for months now, and the released version is quite stable and efficient. It has a lot of additional features compared to the original Panorama View (and of course, it does include a panorama view). 
Number of users is larger than other addons, it is actively developped and reviews are excellent. 
You definitely should give it a try.

**Comment by e-motiv** [comment]
I've been using Simple Tab Groups too, but it's not stable at all! It loses
tabs, urls and messes with Tab session manager. So I sometimes have to
restore tab backups with sometimes some tabs losses since the tab session
backups dosn't always included the latest of them.  But I still keep using
it because there is nothing better out there for the moment. As you said it
has important features that none others offer.

And yeah, I could make a bug report, but I've been doing that on so many
addons, software, etc. I use, I'm just tired of it and have my own addons
to debug and worry about. Sorry...At least some people are warned here
now.. :-/

Cheers,
[image: Ruby Contact Card]
<http://www.e-motiv.net/Ruben%20Goethals%20-%20e-motiv.vcf> Ruben
<Ruben@e-motiv.net>    @   e-motiv <http://www.e-motiv.net>
[image: Ruby] <http://www.e-motiv.net>
+32  (0) 484  /  23 78 78
